### PR TITLE
attach index information when performing status lookup

### LINKED ISSUE #1310: We're not handling added in index + deleted in working directory
As I was updating [mapStatus](https://github.com/desktop/desktop/blob/54b351bb2c94d4b64c1a553bf78515ca46e2b0a4/app/src/lib/git/status.ts#L35-L63) to work with the new porcelain v2 parser in #1305 I noticed that we didn't have a special case for `AD` which is added in index and deleted in work-tree.

This isn't your typical status which is probably why we ignored it in the first place.

```
> touch foo
> git add foo
> rm foo
> git status --porcelain=2
1 AD N... 000000 100644 000000 0000000000000000000000000000000000000000 e69de29bb2d1d6434b8b29ae775ad8c2e48c5391 foo
```

We currently treat this file as modified whereas I think we should omit the file entirely. From the perspective of Desktop we don't care about the index and we'll blow away the index before committing anyway.

image

The inverse scenario is not something we handle well either (a file deleted in the index that's still in the working directory)

```
> touch foo
> git add foo
> git commit -m 'foo'
> git rm --cached foo
> git status --porcelain=2
1 D. N... 100644 000000 000000 e69de29bb2d1d6434b8b29ae775ad8c2e48c5391 0000000000000000000000000000000000000000 foo
? foo
```

This shows up as 2 changed files in desktop

image

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #1310

I've been wrestling with how to tackle #1310 without hacks and finally settled on an approach that I think is better in the long run, and unlocks scenarios like conflict resolution when we want to go down this path.

This PR replaces the existing `mapStatus` function (which would just return an enum based on a number of known scenarios) with a more granular set of data. 

### Why?

GitHub Desktop hasn't always been in sync with how Git represents status changes, and this is mostly around how Git tracks the index state. We've been discarding this information and performing our Git operations with the assumption there is some index state that we need to work around.

There's also some cruft hanging around with the move to the porcelain v2 format. We can't nuke this completely due to `git log` using the v1 status code, but we can isolate that stuff there for now.

### How?

Rather than continue down this path dealing with corner cases, let's split the status into two parts:

 - `GitFileStatus`: how Git represents the status entry
 - `AppFileStatus`: how Desktop should display the status entry (replaces our current `FileStatus` enum)

We can refine this over time, but by introducing these two abstractions we can maintain compatibility in the app and handle edge cases better.

This is what our `FileStatus` now looks:

```ts
export type FileStatus = OrdinaryChange |
  RenamedOrCopiedChange |
  UnmergedChange |
  UntrackedChange
```

This mirrors the structures returned by the v2 porcelain format - see the docs for [git-status](https://git-scm.com/docs/git-status#_changed_tracked_entries) for more context.

I'd like to eventually get this to the point where it all lives in the status parser (there's similar rules there for how each row is parsed that mirror this structure) but this is a good intermediate step down that path.

For each of the types in this union, this gives us some more flexibility with the fields we return. For example, this is how we can represent ordinary changes and renames/copies:

```ts
type OrdinaryChange = {
  readonly kind: 'ordinary'
  readonly type: 'added' | 'modified' | 'deleted',
  /** the staged status of the file (if known) */
  readonly staged?: GitFileStatus,
  /** the unstaged status of the file (if known) */
  readonly unstaged?: GitFileStatus,
}

type RenamedOrCopiedChange = {
  readonly kind: 'renamed' | 'copied',
  /** the staged status of the file (if known) */
  readonly staged?: GitFileStatus,
  /** the unstaged status of the file (if known) */
  readonly unstaged?: GitFileStatus,
}
```

We can make these types more well-defined once we sort out how to handle more of the edge cases (like our fallback value `{ kind: 'ordinary', type: 'modified' }`) and get them under test.

There's also more information from the porcelain status which might be of interest down the track - for example file mode changes, submodule state, rename/copy score - so having a more formal structure for these changes will make this stuff possible.

And for unresolved conflicts we can return a slightly different shape to the data because this no longer reflects the index state:

```ts
type UnmergedChange = {
  readonly kind: 'conflicted',
  /** the first character of the short code ("ours")  */
  readonly us: GitFileStatus,
  /** the second character of the short code ("theirs")  */
  readonly them: GitFileStatus,
}
```

### What?

So what does this unlock? Here's what the rules to fix #1310 now look like:

```
const status = mapStatus(entry.statusCode)

if (status.kind === 'ordinary') {
  if (status.staged === GitFileStatus.Added
      && status.unstaged === GitFileStatus.Deleted) {
    // we can safely avoid drawing this file in the app
    continue
  }
}

if (status.kind === 'untracked') {
  const existingEntry = files.findIndex(p => p.path === entry.path)
  if (existingEntry > -1) {
    // a previous entry for this path exists, remove it and show the untracked file
    files.splice(existingEntry, 1)
  }
}

// for now we just poke at the existing summary
const summary = convertToAppStatus(status)
```

It's not pretty, but becomes more explicit and less hacky than our previous work.

### Performance? Correctness?

I have a test harness to review the performance of this codepath against some large status changes, but the initial proof-of-concept for `mapStatus` was about the same time as the old one. I'll add some more notes to this PR to see how it's affected.

I'll also add some more tests in the follow-up PR to close out #1310 to ensure these edge cases can still be committed by Git and that we aren't affecting users workflows.

### TODO

 - [x] documentation for new abstraction
 - [x] refine tests and confirm we can commit these problematic changes
 - ~~[ ] review `git log` and see if we can simplify the classic `mapStatus` further~~ :boot:

### CODE REVIEW

File: app/src/models/status.ts [hunk 1]
```diff
@@ -2,32 +2,99 @@ import { DiffSelection } from './diff'
...
-  Copied,
 }
 
+/** The porcelain status for an ordinary changed entry */
+type OrdinaryEntry = {
+  readonly kind: 'ordinary'
+   /** how we should represent the file in the application */
+  readonly type: 'added' | 'modified' | 'deleted',
+  /** the staged status of the file (if known) */
+  readonly staged?: GitStatusEntry,
```
Comment (joshaber) [inline app/src/models/status.ts]: `staged` feels like a funny name to me. I tend to think of "staged" as a boolean indicating whether the entry exists in the index.

The `git status` man page seems to use phrases like "status of the index" vs. "status of the work tree." Since this PR is trying to conform more to git, I think our phrasing should too. Thoughts?

File: app/src/lib/git/log.ts [hunk 2]
```diff
@@ -1,9 +1,36 @@
...
+ * shamelessly borrowed from GitHub Desktop (Windows)
+ */
+function mapStatus(rawStatus: string): AppFileStatus {
+
+  // TODO: This is due to the fact that porcelain V2 changed from
+  // using space to using a dot when either side is unmodified.
+  // We should probably parse this properly. We still trim the space
+  // since mapStatus is used from log.ts as well which passes
+  // porcelain v1 status codes.
+  const status = rawStatus.replace(/[ .]/, '')
```
Comment (joshaber) [inline app/src/lib/git/log.ts]: If this `mapStatus` is used only for `log`, we can remove this, can't we?

File: app/src/lib/git/status.ts [hunk 3]
```diff
@@ -75,9 +67,28 @@ export async function getStatus(repository: Repository): Promise<IStatusResult>
...
+            && status.unstaged === GitStatusEntry.Deleted) {
+          // we can safely avoid drawing this file in the app
+          continue
+        }
+      }
+
+      if (status.kind === 'untracked') {
+        const existingEntry = files.findIndex(p => p.path === entry.path)
+        if (existingEntry > -1) {
+          // a previous entry for this path exists, remove it
```
Comment (joshaber) [inline app/src/lib/git/status.ts]: When would this happen?

File: app/src/lib/git/status.ts [hunk 4]
```diff
@@ -75,9 +67,28 @@ export async function getStatus(repository: Repository): Promise<IStatusResult>
   for (const entry of parsePorcelainStatus(result.stdout)) {
     if (entry.kind === 'entry') {
       const status = mapStatus(entry.statusCode)
+
+      if (status.kind === 'ordinary') {
+        if (status.staged === GitStatusEntry.Added
+            && status.unstaged === GitStatusEntry.Deleted) {
+          // we can safely avoid drawing this file in the app
```
Comment (joshaber) [inline app/src/lib/git/status.ts]: So this is the case where a file has been staged but no longer exists on disk, is that right? So it won't really end up in the commit?

File: app/src/lib/status-parser.ts [hunk 5]
```diff
@@ -137,3 +139,203 @@ function parseUntrackedEntry(field: string): IStatusEntry {
     path,
   }
 }
+
+
+/**
+ * Map the raw status text from Git to a structure we can work with in the app.
+ *
+ * This relies on the porcelain v2 format and status codes, so for
+ * interoperability the existing v1 code is still around for now.
```
Comment (joshaber) [inline app/src/lib/status-parser.ts]: Can we 🔥 this comment about v1 code, since that's encapsulated in `log` now?

File: app/src/models/status.ts [hunk 6]
```diff
@@ -2,32 +2,99 @@ import { DiffSelection } from './diff'
...
 
-/** the state of the changed file in the working directory */
-export enum FileStatus {
+/**
+ * The status entry code as reported by Git.
+ *
+ * NOTE: 'U' is considered "Updated but unmerged", and is a valid code here,
+ * but instead we mark this as "Modified" when surfacing conflicts.
+ *
+ * TODO (BF): what as I thinking here? Tidy up this comment or remove it.
```
Comment (joshaber) [inline app/src/models/status.ts]: LOL

File: app/src/lib/git/status.ts (hunk 4 above)
Comment (shiftkey) [inline app/src/lib/git/status.ts]: It depends on whether the content of the untracked file with the same path has changed:

 - no content change - no-op
 - content change - modified

I'd need to look at what diff we render here at this stage. It could be confusing.

File: app/src/lib/git/log.ts (hunk 2 above)
Comment (shiftkey) [inline app/src/lib/git/log.ts]: I think so? Let's find out!

File: app/src/models/status.ts (hunk 1 above)
Comment (shiftkey) [inline app/src/models/status.ts]: > The `git status` man page seems to use phrases like "status of the index" vs. "status of the work tree.

Ooooh this is good. :metal:

File: app/src/lib/git/status.ts (hunk 3 above)
Comment (shiftkey) [inline app/src/lib/git/status.ts]: `git status` will list the untracked files after all the tracked files. This relates to the point above about the "staged delete and untracked file with same path" case.

**Comment by shiftkey** [comment]
:apple:

File: app/src/lib/git/status.ts [hunk 7]
```diff
@@ -75,9 +67,28 @@ export async function getStatus(repository: Repository): Promise<IStatusResult>
   for (const entry of parsePorcelainStatus(result.stdout)) {
     if (entry.kind === 'entry') {
       const status = mapStatus(entry.statusCode)
+
+      if (status.kind === 'ordinary') {
+        if (status.index === GitStatusEntry.Added
```
Comment (joshaber) [inline app/src/lib/git/status.ts]: Could we add a comment here about why we ignore these types of changes?

File: app/src/lib/git/status.ts (hunk 3 above)
Comment (joshaber) [inline app/src/lib/git/status.ts]: Could we document these cases so we know why this can happen?

**Comment by shiftkey** [comment]
:apple:

**Comment by joshaber** [comment]
It looks like Appveyor wants a few more config fixes?

**Comment by shiftkey** [comment]
🍎 🍎
